### PR TITLE
Feature/591 Diff Viewer Improvements

### PR DESCRIPTION
**Pull request by bruderol**
Proposed changes by @til77 to be reviewed and further discussed how to deal and proceed with them.

Demo for this feature branch can be found here: http://demo.scenarioo.org/scenarioo-diff-viewer-plus 

See also issue #591

### CODE REVIEW

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java [hunk 1]
```diff
@@ -18,167 +18,149 @@
...
+			// convert images to ARGB format to ensure that the data buffers store one pixel as ARGB
+			BufferedImage oldImageArgb = new BufferedImage(unifiedWidth, unifiedHeight, BufferedImage.TYPE_INT_ARGB);
+			BufferedImage newImageArgb = new BufferedImage(unifiedWidth, unifiedHeight, BufferedImage.TYPE_INT_ARGB);
+
+			// for simplicity paint both images on the upper left corner of the compare image since searching the sub-image is too expensive
+			// areas not covered by images are transparent which will lead to a maximum difference on all 3 measured color channels
+			oldImageArgb.getGraphics().drawImage(oldImage, 0, 0, null);
+			newImageArgb.getGraphics().drawImage(newImage, 0, 0, null);
+
+			BufferedImage diffImage = new BufferedImage(Math.max(oldImageArgb.getWidth(), newImageArgb.getWidth()), Math.max(oldImageArgb.getHeight(), oldImageArgb.getHeight()), BufferedImage.TYPE_INT_ARGB);
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Why oldImage.getWidth() should be different than oldImageArgb.getWidth()? Image size calculation already done in line 97.

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java [hunk 2]
```diff
@@ -18,167 +18,149 @@
...
+			double absoluteErrorTotal = 0;
+
+			// improve performance by accessing data buffers directly since methods to access pixels through BufferedImage are passed through the ColorModel which slows things down
+			DataBuffer oldImageBuffer = oldImageArgb.getRaster().getDataBuffer();
+			DataBuffer newImageBuffer = newImageArgb.getRaster().getDataBuffer();
+			DataBuffer diffImageBuffer = diffImage.getRaster().getDataBuffer();
+
+			int totalPixels = oldImageBuffer.getSize();
+
+			for (int n = 0; n < oldImageBuffer.getSize(); n++) {
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: use "totalPixels"

Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Maybe could use Java 8 stream api. Doing this work with parallelStream could improve the performance.

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java [hunk 3]
```diff
@@ -18,167 +18,149 @@
...
-	}
-
 	/**
-	 * A Diff Screenshoot will be created and stored in the directory of the diffScreenshot path.
+	 * A Diff Screenshot will be created and stored in the directory of the diffScreenshot path.
 	 * This directory, and all the parent directories, will be created if they do not exist.
 	 */
-	double compareScreenshots(final File baseScreenshot, final File comparisonScreenshot,
-			final File diffScreenshot) {
+	double compareScreenshots(final File baseScreenshot,
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: This method is doing more things than compoareScreenshots. It's more about compareCreateRequiredDirectoriesForScreenshotAndStore method. Some Clean Code may help

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java [hunk 4]
```diff
@@ -18,167 +18,149 @@
...
-			if (m.find()) {
-				return Math.sqrt(Double.parseDouble(m.group(0))) * 100;
+
+			double mae = absoluteErrorTotal / totalPixels;
+
+			if (diffAvailable) {
+				ImageIO.write(diffImage, "PNG", diffScreenshot);
 			}
+
+			return Math.sqrt(mae) * 100;
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Magic code

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement
+@XmlAccessorType(XmlAccessType.FIELD)
+public class ComparisonResult {
+
+	private BuildIdentifier baseBuild;
+	private BuildIdentifier compareBuild;
+
+	ComparisonConfiguration comparisonConfiguration;
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: Why there is no private modifier?

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+	public void setComparisonConfiguration(ComparisonConfiguration comparisonConfiguration) {
+		this.comparisonConfiguration = comparisonConfiguration;
+	}
+
+	public String getRmaePercentage() {
+		return rmaePercentage;
+	}
+
+	public void setRmaePercentage(double rmaePercentage) {
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: This should be a string. Because getter and setter of a member should be the same type.

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/DiffViewerStatus.java [hunk 7]
```diff
@@ -4,13 +4,4 @@
 
 @XmlRootElement
 public class DiffViewerStatus {
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/DiffViewerStatus.java]: Empty class. Should be deleted

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java [hunk 8]
```diff
@@ -52,7 +51,6 @@ public ApplicationStatus getApplicationStatus() {
 		ApplicationStatus applicationStatus = new ApplicationStatus();
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java]: This class has Constructor and getter/setter which are never used.

File: scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java [hunk 9]
```diff
@@ -98,6 +95,35 @@ public void reimportBuild(@PathParam("branchName") final String branchName,
 		ScenarioDocuBuildsManager.INSTANCE.reimportBuild(buildIdentifier);
 	}
 
+	@GET
+	@Path("importBuild/{branchName}/{buildName}/{comparisonBranchName}/{comparisonBuildName}/{comparisonName}")
+	@Produces({"application/json"})
+	public Response importBuild(
```
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: No tests available

**Comment by kabl** [comment]
Not sure but I guess this documentation needs an update: https://github.com/scenarioo/scenarioo/blob/develop/docs/features/diff-viewer/setup.md

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 4 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Indeed - magic - Think we should remove that - @til77 what do you think?

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 2 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Cool Idea - Open an issue for it - if you have time you could do that.

Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: @kabl not understand your comment?

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 1 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: yes that looks strange - think we can clean that up

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 3 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: Maybe not the best name, agree, but your name is not a better name for me - sorry - the proposed alternative would be less understandable for me to quickly understand what this method is doing - maybe something like `compareScreenshotAndStoreDiffImage` - creating parent directories for storing something is something we always do, whenever it is needed for storing something. So that is not realy an unexpected side effect here.

Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: by the way - also it is somehow clear from the context, that all the methods here just store their results in the file system, that is why I think the original name is not so bad, as you may think.

@danielsuter What do you think?

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java [hunk 10]
```diff
@@ -18,167 +18,149 @@
...
+		} catch (IOException ex) {
+			LOGGER.warn("Failed to compare images (base: " + baseScreenshot + ", comparison: " + comparisonScreenshot + ")", ex);
+			return SCREENSHOT_DEFAULT_CHANGE_RATE;
```
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: no - I think not - can explain it to you.

Short answer: Since this is a background job the log is stored in a log file, that the user could browse (either in file system on the server or later through a special view in the web app, which is currently missing --> there is an issue for that)

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java (hunk 5 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: Dont know, is it used somewhere? Maybe just lazy code?

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/DiffViewerStatus.java (hunk 7 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/DiffViewerStatus.java]: yes, indeed - well spotted! --> that will also have impact on the client side, because there is some config view where this status is displayed - that is dead code that has to be removed everywhere it is still used (in REST service and as well on the client side).

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java (hunk 8 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java]: which class?

**Comment by bruderol** [comment]
@kabl: concerning last comment: yes, sure - there is an issue for that.

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 2 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: ah, yes!

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java (hunk 6 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: why is it a string? why dont we store that as a number? @til77 can you clarify on that?

File: scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java (hunk 9 above)
Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: oh - i see - that one looks like we need to discuss it further - I not like the REST service to wait for potentially endless unless some background tasks are done. We have to discuss a different solution here. @til77  wouldnt it be also okay for you if we simply provide another service that checks for a comparison result and returns nothing if it is not available yet and then let the build poll for that information until it is available?

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 3 above)
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: My method name was not a proposal. It's more about split this method in smaller parts. Or move some calculation in methods.

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java (hunk 8 above)
Comment (kabl) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/ConfigurationResource.java]: Sorry. The class was not modified in this PR. It's ```ApplicationStatus.java```

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 1 above)
Comment (til77) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: yes, can be cleaned up by using unifiedWidth and unifiedHeight

File: scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java (hunk 4 above)
Comment (til77) [inline scenarioo-server/src/main/java/org/scenarioo/business/diffViewer/comparator/ScreenshotComparator.java]: I have no idea why the sqrt is used. Maybe it was used to emphasize the difference on very small numbers like a diff of 0.0001 to reach one percent change. Our change to always round differences up to 1% maybe makes this calculation obsolete?

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java (hunk 5 above)
Comment (til77) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: I assume that this was missed.

File: scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java (hunk 6 above)
Comment (til77) [inline scenarioo-server/src/main/java/org/scenarioo/model/diffViewer/ComparisonResult.java]: I see no reason for the type mess. Change it to double if the xml libraries have no trouble with double.

File: scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java (hunk 9 above)
Comment (til77) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: Changing to async is also fine for us however the importBuild method would have to return a reference ID so that the client can ask about the status for this particular import.

Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: @til77 yes, sure, there is already such an "id" and this is implicitly given by "branch name" + "build name" + the configured "comparison name" - and there is a also a REST service to get the result once it is calculated. Would that be enough?

Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: @til77 the URL to get the diff info for a comparison, is: http://demo.scenarioo.org/scenarioo-diff-viewer-plus/rest/diffViewer/baseBranchName/{put-your-branch-name-here}/baseBuildName/{put-your-build-name-here}/comparisonName/{put-the-name-of-the-comparison-here}/buildDiffInfo  -returned json contains the result in field "changeRate" as a decimal number

Comment (bruderol) [inline scenarioo-server/src/main/java/org/scenarioo/rest/builds/BuildsImporterResource.java]: @kabl @til77 I specified in issue #605 what I think we will have to change.

File: scenarioo-docu-generation-example/.gitignore [hunk 11]
```diff
@@ -1,4 +1,5 @@
 build
+out
```
Comment (bruderol) [inline scenarioo-docu-generation-example/.gitignore]: @adiherzog Why is this needed now?? I not understand it

File: scenarioo-docu-generation-example/src/test/resources/example/documentation/scenarioDocuExample/wikipedia-docu-example-dev/2014-04-19/build.xml [hunk 12]
```diff
@@ -3,5 +3,6 @@
     <name>2014-04-19</name>
     <revision>F398DA3</revision>
     <date>2014-04-19T00:00:00+02:00</date>
+    <status>success</status>
```
Comment (bruderol) [inline scenarioo-docu-generation-example/src/test/resources/example/documentation/scenarioDocuExample/wikipedia-docu-example-dev/2014-04-19/build.xml]: @adiherzog This commit to change the test data was unnecessary - the example documentaiton data should always be commited (only!) after a clean run of "scenarioo-docu-generation-example" - this fields are added by the viewer after importing the builds the first time, and it is intended to not have them in the example data for testing that the state gets automatically added by the server. But yes, I know, this is not optimal and we might have to improve that once. until we improve it we should only commit example data, when generated with a clean build.

Comment (bruderol) [reply scenarioo-docu-generation-example/src/test/resources/example/documentation/scenarioDocuExample/wikipedia-docu-example-dev/2014-04-19/build.xml]: @adiherzog hmmm - ok, forget - it seems the tests are now generating that status for the builds always??? Why??? I think this has been changed wrongly.

Comment (bruderol) [reply scenarioo-docu-generation-example/src/test/resources/example/documentation/scenarioDocuExample/wikipedia-docu-example-dev/2014-04-19/build.xml]: @adiherzog: I do not get it - somehwo it seems that you copied it from the "out" directory - there it is generated with the state - I not get why there is this "out"-directory - cant we change that to write to build as well and use meaningful name why this output is produced (probably testing the import)

Comment (bruderol) [reply scenarioo-docu-generation-example/src/test/resources/example/documentation/scenarioDocuExample/wikipedia-docu-example-dev/2014-04-19/build.xml]: @adiherzog ahhhrgh - okay - think now I understand - sorry - my first comment was right - the version in the out directory is always the commited one (gets copied during build somehow) - will now copy again the version from build and then it should be okay again.

File: scenarioo-docu-generation-example/.gitignore (hunk 11 above)
Comment (adiherzog) [reply scenarioo-docu-generation-example/.gitignore]: I had a wrong IntelliJ configuration that compiled into the out folder. So I will remove "out" again.

File: scenarioo-docu-generation-example/.gitignore [hunk 13]
```diff
@@ -1,5 +1,4 @@
 build
-out
```
Comment (bruderol) [inline scenarioo-docu-generation-example/.gitignore]: hmm, sorry, maybe my comment about this was bad, I have seen that I have this 'out' as well - so just removing it from gitignore does not realy solve the problem - either put it back or adjust that everything is built into `build` directory

**Comment by bruderol** [comment]
@adiherzog and me decided to merge that now -  we will continue to improve diff viewer further on additional smaller feature branches
